zipkin-autoconfigure/storage-cassandra3/src/main/java/zipkin/autoconfigure/storage/cassandra3/brave/TracedSession.java

```diff
@@ -11,204 +11,182 @@
+        useKeyspace(session, statement.getKeyspace());
+        return super.prepareAsync(statement);
     }
 
-    @Override public ResultSet getUninterruptibly(long timeout, TimeUnit unit)
-        throws TimeoutException {
-      return delegate.getUninterruptibly(timeout, unit);
+    @Override
+    public CloseFuture closeAsync() {
+        return session.closeAsync();
     }
 
-    @Override public boolean cancel(boolean mayInterruptIfRunning) {
-      return delegate.cancel(mayInterruptIfRunning);
+    @Override
+    public boolean isClosed() {
+        return session.isClosed();
     }
 
-    @Override public boolean isCancelled() {
-      return delegate.isCancelled();
+    @Override
+    public Cluster getCluster() {
+        return session.getCluster();
     }
 
-    @Override public boolean isDone() {
-      return delegate.isDone();
+    @Override
+    public Session.State getState() {
+        return session.getState();
     }
 
-    @Override public ResultSet get() throws InterruptedException, ExecutionException {
-      return delegate.get();
+    private void useKeyspace(final Session session, final String ks) {
+        String keyspace = null != ks ? ks : session.getLoggedKeyspace();
+        // this method exists because ZipkinTracingCluster.connect() loops if it calls super.connect(keyspace)
+        if (!keyspaceSet) {
+            try {
+                ResultSetFuture future = session.executeAsync("USE " + QueryBuilder.quote(keyspace));
+                Uninterruptibles.getUninterruptibly(future, useKeyspaceTimeout, TimeUnit.MILLISECONDS);
+                keyspaceSet = true;
+            } catch (TimeoutException e) {
+                throw new DriverInternalError(""
+                        + "No responses after " + useKeyspaceTimeout + " milliseconds while setting current keyspace. "
+                        + "This should not happen, unless you have setup a very low connection timeout.");
+            } catch (ExecutionException e) {
+                // see DefaultResultSetFuture.extractCauseFromExecutionException(e);
+                if (e.getCause() instanceof DriverException) {
+                    throw ((DriverException) e.getCause()).copy();
+                } else {
+                    throw new DriverInternalError("Unexpected exception thrown", e.getCause());
+                }
+            } catch (RuntimeException e) {
+                session.close();
+                throw e;
+            }
+        }
     }
 
-    @Override public ResultSet get(long timeout, TimeUnit unit)
-        throws InterruptedException, ExecutionException, TimeoutException {
-      return delegate.get(timeout, unit);
+    @Override
+    protected ListenableFuture<PreparedStatement> prepareAsync(String query, Map<String, ByteBuffer> customPayload) {
+        return session.prepareAsync(query);
     }
 
-    @Override public void addListener(Runnable listener, Executor executor) {
```

Comment:
I think the reason I did this (created a wrapped future vs chaining) was I a heck of a time getting async to actually work properly. I presume you've verified this patch actually works as expected (versions have changed, since so maybe a driver bug I hit was been squashed)